### PR TITLE
clippy improvements to iterators

### PR DESCRIPTION
**Pull request by llogiq**
Nothing too big, a few needless returns and a few closures eliminated (the latter may improve performance in some cases, at least compilation should be a bit faster).

### CODE REVIEW

File: src/libcore/iter.rs [hunk 1]
```diff
@@ -1698,7 +1698,7 @@ impl<B, I: DoubleEndedIterator, F> DoubleEndedIterator for Map<I, F> where
 {
     #[inline]
     fn next_back(&mut self) -> Option<B> {
-        self.iter.next_back().map(|a| (self.f)(a))
+        self.iter.next_back().map(self.f)
```
Comment (eddyb) [inline src/libcore/iter.rs]: How does this work? Closures are not `Copy`, and you only have mutable access, can't move out of `self` - did you mean to write `&mut self.f`?

**Comment by Manishearth** [comment]
r+ if it compiles, else use `&mut self.f` there

**Comment by llogiq** [comment]
Strangely it passed a build on my box. @Manishearth you are right, we need `&mut  self.f`.

**Comment by llogiq** [comment]
@Manishearth Does homu need another r+ to run?

**Comment by Manishearth** [comment]
@xrpo p}

**Comment by Manishearth** [comment]
@bors r+ 

welp, dvorak

**Comment by llogiq** [comment]
:smile:
Am 03.09.2015 17:51 schrieb Manish Goregaokar <notifications@github.com>:@bors r+ 

welp, dvorak

**Pull request merged**